### ISSUE TITLE
`this.path()` not working in RStudio, incorrectly claims R is running in an unrecognized manner

### ISSUE THREAD

**Issue by CorneeldH** [ArcadeAntics/this.path#7]
Thanks for the package, started using it for some fun internal functions (clear_script_objects function). However, when I use version 1.0.x or higher I get the following error when I run it in R Studio:
```
Error in .this.path(verbose, original, for.msg) : 
  'this.path' used in an inappropriate fashion
* no appropriate source call was found up the calling stack
* R is being run in an unrecognized manner
 ```

I think the issue is in the fact it doesn't recognize it is within R Studio. Source does work fine. 

I tried looking at the source code, but didn't immediately understood the gui.rstudio object. When I slighty adapt your answer here: https://stackoverflow.com/a/64129649 (fixing only adc into context on line 143 it does work). Also when I downgrade to version 0.8.0 (I had that installed in another project) it works.

R Studio version: 2022.07.2 Build 576. I work in projects with .Rproj files, on PC with Windows 11. 

For now I've just set it to 0.8.0 in my renv, but if you have a clue or if I can  test something let me know!

**Comment by ArcadeAntics** [comment]
The `gui.rstudio` object is a promise, you can find it in `/this.path/R/promises.R` or here: https://github.com/ArcadeAntics/this.path/blob/main/R/promises.R

The idea is that I have a bunch of conditions I want to evaluate to determine how R is being run, but that each condition should remain constant in any given session. So everywhere you see `gui.rstudio`, it's like running `.Platform$GUI == "RStudio"`. The advantages are that it's faster to evaluate only once and that if I discover a faster / / better way of determining how R is running, I only have to change `/this.path/R/promises.R` instead of looking through the files and correcting it everywhere.

And thanks for pointing out the problem at https://stackoverflow.com/a/64129649, I tried renaming some variables to have it make more sense but I must have missed that correction.

Can you try running `.Platform` or `.Platform$GUI` in your RStudio session right before calling `this.path()`. The name of the GUI should be `"RStudio"`, but RStudio renamed themselves to Posit, so they may have changed the value of the string as well.

**Comment by CorneeldH** [comment]
Thanks for the quick response and explanation! Our data pipeline is pretty straightforward so I hadn't used promises yet. Nice to know they exist. 

No luck with the platform though, it is still RStudio:

![afbeelding](https://user-images.githubusercontent.com/6919390/207001338-1cb128eb-d998-4a36-993d-71828eabeade.png)

Manual running of promise:

![afbeelding](https://user-images.githubusercontent.com/6919390/207001805-87789c4e-f082-427b-801c-a5167aaea7c3.png)

Some more info:
```
R version 4.2.1 (2022-06-23 ucrt) -- "Funny-Looking Kid"
Copyright (C) 2022 The R Foundation for Statistical Computing
Platform: x86_64-w64-mingw32/x64 (64-bit)
```

I can check which this.path version 'breaks' it (0.8.0 works, 1.0.x (can't remember the x) doesn't)?

**Comment by ArcadeAntics** [comment]
I think I need the version of this.path that breaks. You can retrieve it with `getNamespaceVersion("this.path")`. I don't remember which version it was, it was around `1.0.2.84`, but I'd made a mistake in that version that got corrected later.

R has a byte compiler for packages, it makes the functions a faster by doing some substitutions and simplifications. I made a promise object in /R/promises.R that was being evaluated early at INSTALL time instead of at load time. This meant that the package thought you were never in RStudio because it evaluated that promise too early.

This might not be the issue, I thought I never released that version of this.path, I thought I fixed it before pushing to GitHub, but I might be wrong. Can you try installing this.path 1.1.0 from CRAN and see if the issue persists? Also, once you've installed that version, can you try running:

```
evalq(PRINFO(gui.rstudio), getNamespace("this.path"))
```

This should tell you more information about the promise object `gui.rstudio`, most importantly it should not have a field named "PRVALUE". If it does not (initially) have that field, then I would expect it to work. Does using the newest version fix this?





Are you using `this.path()` in your startup code? In a `Rprofile.site` or `.Rprofile` file? There's a weird situation where `.Platform$GUI` is set to `"Rgui"` on startup, then changed to `"RStudio"` after startup, right before you're given access to the R Console. I'll see if there's anything I can do about that.

**Comment by CorneeldH** [comment]
The newest didn't work, but I re-installed it for testing the evalq code:

![afbeelding](https://user-images.githubusercontent.com/6919390/207300099-3c09a5cd-069d-4ed5-a0ed-b01be9cfd70b.png)

It does have PRVALUE on FALSE. 

I have tried with 1.1.0 , 0.8.0 (and I think also with 1.0.x). From >=1 it didn't work

**Comment by ArcadeAntics** [comment]
Can you email me your Discord or Facebook contact info so we can video chat and share screens, or maybe we can use gmail's chat feature? I don't think I have enough information to solve the issue as it is right now. My email is `ikwsimmo@gmail.com`. I'm available from 2100 - 1300 UTC. I'll summarize whatever we talk about here so that anyone else having your issue can see how it was resolved.

In the meantime, can you restart your computer and install the new build I just put on github?:

```
install.packages("this.path", repos = "https://raw.githubusercontent.com/ArcadeAntics/PACKAGES")
```

**Comment by CorneeldH** [comment]
Cool, this seems to work!

```
> install.packages("this.path", repos = "https://raw.githubusercontent.com/ArcadeAntics/PACKAGES")
Installing package into ‘C:/Users/user/LocalDocuments/R/x86_64-w64-mingw32-library/4.2’
(as ‘lib’ is unspecified)
trying URL 'https://raw.githubusercontent.com/ArcadeAntics/PACKAGES/bin/windows/contrib/4.2/this.path_1.1.0.4.zip'
Content type 'application/zip' length 260596 bytes (254 KB)
downloaded 254 KB

package ‘this.path’ successfully unpacked and MD5 sums checked

The downloaded binary packages are in
	C:\Users\user\AppData\Local\Temp\RtmpaW1sL7\downloaded_packages
evaluating 'initwd'
> this.path::this.path()
evaluating 'gui.rstudio'
evaluating 'debugSource'
evaluating 'tools.rstudio'
evaluating 'in.shell'
evaluating 'os.unix.in.shell'
evaluating 'os.unix'
evaluating 'os.windows.in.shell'
evaluating 'os.windows'
evaluating '.rs.api.getActiveDocumentContext'
[1] "C:/Users/user/LocalDocuments/repositories/test_without_renv/test_this_path.R"
Warning message:
In fun(libname, pkgname) : this version is only for debugging purposes
 go to the archive and install the previous version
 ```

**Comment by ArcadeAntics** [comment]
Leaving this issue open until the update has been published on CRAN.

**Comment by ArcadeAntics** [comment]
Hi again @CorneeldH 

Can you please install the newest update:

```R
install.packages("this.path", repos = "https://raw.githubusercontent.com/ArcadeAntics/PACKAGES")
```

and confirm that `this.path()` works in RStudio? If so, then I can publish the update to CRAN. Thank you!

**Comment by CorneeldH** [comment]
It works great, cheers!

**Comment by ArcadeAntics** [comment]
The update 1.2.0 has been uploaded to CRAN, the binaries have been made and will be available shortly.

If you have any other problems, please open another issue. Thank you for using my R package!
